File: Question1/student_record.py
grade_points = {"A": 4.0, "B": 3.0, "C": 2.0, "D": 1.0}

student_list = [
    {
        "student_id": "S001",
        "courses": [
            {"name": "Backend Development", "credits": 4, "grade": "A"},
            {"name": "DAS", "credits": 4, "grade": "B"}
        ]
    }
]

def student_report(student_id):
    for student in student_list:
        if student["student_id"] == student_id:
            return student
    return None

def show_schedule(student):
    print('\nCourse Schedule:')
    for c in student["courses"]:
        print(f"{c['name']} - Credits: {c['credits']}")

def add_grade(student, course_name, grade, credits=3):
    for c in student["courses"]:
        if c["name"].lower() == course_name.lower():
            c["grade"] = grade
            print(f"Grade for {course_name} updated to {grade}.")
            return

    student["courses"].append({"name": course_name, "credits": credits, "grade": grade})
    print(f"Course '{course_name}' added with grade {grade} and {credits} credits.")

def calculate_gpa(student):
    total_points = 0
    total_credits = 0
    for c in student["courses"]:
        if c.get("grade") in grade_points:
            total_points += grade_points[c["grade"]] * c["credits"]
            total_credits += c["credits"]
    if total_credits == 0:
        print("No grades yet.")
    else:
        print(f"GPA: {total_points/total_credits:.2f}")

def show_progress(student):
    print('\nStudent Progress:')
    for c in student["courses"]:
        grade = c.get("grade", "Not graded")
        print(f"{c['name']}: {grade}")



student_id = "S001"
student = student_report(student_id)
if not student:
    print("Student not found.")
else:
    show_schedule(student)
    add_grade(student, "New Course", "A", credits=2) 
    add_grade(student, "Backend Development", "B")   
    show_progress(student)
    calculate_gpa(student)

File: Question1/student_record.js
const gradePoints = { A: 4.0, B: 3.0, C: 2.0, D: 1.0, F: 0.0 };
let studentRecord = [
  {
    studentId:4545,
    name:"Qefar",
    courses: [
      { name: "Math", instructor: "Dr. Katya", creditscore: 3 },
      { name: "History", instructor: "Mr Joseph", creditscore: 2 }
    ],

    studentId:5656,
    name:"jomo",
    courses:[
        {name:"javascript",instructor:"Dr.Keshy",creditscore:3},
        
    ]

    
  }
  
];

function findStudent(studentId) {
  return studentRecord.find(student => student.studentId === studentId);
}

function viewCourseSchedule(studentId) {
  for (student in studentRecord){
    const student = findStudent(studentId);
  
  if (!student) {
    console.log("Student not found.");
    return;
  }
  console.log(`Schedule for ${student.name}:`);
  student.courses.forEach(course => {
    console.log(`${course.name} - Instructor: ${course.instructor}, Credits: ${course.creditscore}`);
  });
}
}

function enterNewGrade(name, courseName, grade) {
  const student = findStudent(name);
  if (!student) {
    console.log("Student not found.");
    return;
  }
  const course = student.courses.find(c => c.name === courseName);
  if (course) {
    course.grade = grade;
    console.log("Grade updated.");
  } else {
    console.log("Course not found.");
  }
}

function calculateGPA(name) {
  const student = findStudent(name);
  if (!student) {
    console.log("Student not found.");
    return;
  }
  let totalPoints = 0;
  let totalCredits = 0;
  student.courses.forEach(course => {
    if (course.grade) {
      const points = gradePoints[course.grade] || 0;
      totalPoints += points * course.creditscore;
      totalCredits += course.creditscore;
    }
  });
  if (totalCredits === 0) {
    console.log("No grades available.");
  } else {
    console.log("GPA:", (totalPoints / totalCredits).toFixed(2));
  }
}

function showProgress(name) {
  const student = findStudent(name);
  if (!student) {
    console.log("Student not found.");
    return;
  }
  console.log(`Progress for ${student.name}:`);
  student.courses.forEach(course => {
    console.log(`${course.name} - Grade: ${course.grade || "Not graded yet"}`);
  });
}


viewCourseSchedule(5656);
enterNewGrade(5656, "javascript", "A");
calculateGPA(5656);
showProgress(5656);
